macro/src/derive.rs

```diff
@@ -0,0 +1,361 @@
+                        #to_java_body
+                    }
+                }
+            }
+
+            impl duchess::plumbing::ToJavaImpl<#class_name> for #self_ty {
+                fn to_java_impl<'jvm>(rust: &Self, jvm: &mut duchess::Jvm<'jvm>) -> duchess::Result<'jvm, Option<Local<'jvm, #class_name>>> {
+                    Ok(Some(duchess::JvmOp::execute_with(rust, jvm)?))
+                }
+            }
+        ))
+    }
+
+    /// Generates the code to create this variant as part of a `ToRust` impl.
+    /// Assumes `self` is the java type and `jvm` is in scope.
+    fn variant_to_rust(
+        &mut self,
+        variant: &VariantInfo,
+    ) -> Result<proc_macro2::TokenStream, syn::Error> {
+        // For each field, construct an expression we will use to initialize its value.
+        let mut initializers = VecDeque::new();
+        for field in variant.ast().fields {
+            if let Some(name) = &field.ident {
+                if name == "this" {
+                    // Special case for fields named this
+                    initializers
+                        .push_back(quote_spanned!(name.span() => self.global().execute_with(jvm)?));
+                } else if self.is_option(&field.ty) {
+                    initializers.push_back(quote_spanned!(name.span() =>
+                    self
+                        .#name()
+                        .to_rust()
+                        .execute_with(jvm)?
+                    ));
+                } else {
+                    initializers.push_back(quote_spanned!(name.span() =>
+                    self
+                        .#name()
+                        .assert_not_null()
+                        .to_rust()
+                        .execute_with(jvm)?
+                    ));
+                }
+            } else {
+                // FIXME: We should probably support something like
+                // `#[duchess::args(foo, bar, bar)]` ?
+                return Err(syn::Error::new(
+                    field.span(),
+                    "tuple structs not yet supported",
+                ));
+            }
+        }
+
+        let mut counter = 0;
+        let construct = variant.construct(|_field, index| {
+            assert!(counter == index);
+            counter += 1;
+            initializers.pop_front().unwrap()
+        });
+
+        Ok(construct)
+    }
+
+    /// Generates the code to create this variant as part of a `ToRust` impl.
+    /// Assumes `self` is the java type and `jvm` is in scope.
+    fn variant_to_java(
+        &mut self,
+        variant: &VariantInfo,
+    ) -> Result<proc_macro2::TokenStream, syn::Error> {
+        let variant_span = variant.ast().ident.span();
+
+        // If there is a field named `this`, just return that.
+        if let Some(binding) = variant
+            .bindings()
+            .iter()
+            .find(|b| b.ast().ident.as_ref().map(|i| i == "this").unwrap_or(false))
+        {
```

Comment:
What does it mean to modify a field in Rust but then use `this` in Java land? It could lead to some unexpected behavior like
```rust
let mut value = compute_something_in_java();
assert_eq!(value.int_field, 2);
value.int_field = 4;
compute_something_else_in_java(&value); // is int_field 2 or 4 from Java's perspective?
```

I'd suggest if we have `this` field, we prevent mutable access to the fields through some getter fns.

Reply:
Hmm. That's true. We could error if there are public fields and a `this` field, but of course it won't prevent people from doing the wrong thing. The other option (more drastic) would be to use a wrapper type that prevents direct writes and only gives `&` access. That seems a bit extreme to me. 

(In retrospect, I wish that Rust had you declare mutable fields as mutable... maybe worth proposing as an RFC sometime.)